[1.15] Adding Categories within JigsawPattern and option to add/remove buildings

Cheaterpaul opened the pull request:
(while there is no option to add buildings to a village)

2 events are fired:

- JigsawPatternInitEvent - fired in constructor of JigsawPattern. There you can add/remove JigsawPieces

- StructureJigsawPoolInitEvent - fired when a jigsaw based structure inititizes its JigsawPatterns. There you can add your own JigsawPattern before the structure need them.

Problem:
when adding more houses to a village chances of spawning e.g. a farm are shrinking.

Resolved by adding JigsawCategories:
the categories preserves the weight of the type of JigsawPieces inside it.

The JigsawPatternInitEvent first creates JigsawCategories from the JigsawPattern's weighted JigsawPiece based on the pieces registryname and groups the pieces inside the category.

Now when subscribed to the JigsawPatternInitEvent you can add/remove JigsawPieced based on there registry from a category or add/remove whole categories from the JigsawPattern. The changes are cached before there are applied all together to the pre created JigsawCategories of the JigsawPattern.


optionally i added a registrylike console output for all changes (JigsawPatternInitEvent#printChanges())

old PR's #6324 (same but for 1.14), #6142 (same without the whole categories for 1.14)

Cheaterpaul committed:
adding feature to add JigsawPieces to Village Pools

Cheaterpaul committed:
adding feature to add JigsawPieces to Village Pools

Cheaterpaul committed:
some modifications

Cheaterpaul committed:
some modifications

Cheaterpaul committed:
rearranging stuff

Cheaterpaul committed:
now able to remove and JigsawPattern for better compatibility,
fixed some issues,
updated test

Cheaterpaul committed:
fixed patch imports

Cheaterpaul committed:
requested changes

Cheaterpaul committed:
fix

Cheaterpaul committed:
moved removeBuilding/addBuilding,
make JigsawPiece lists public

Cheaterpaul committed:
now fires two events:
one on jigsawpattern constructor to directly add/remove all pices,
one on Village/PillageOutpost to register jigsawpattern before the structure is constructed

Cheaterpaul committed:
fix merge conflicts

Cheaterpaul committed:
edit comment

Cheaterpaul committed:
village fired attribute for single event firing

Cheaterpaul committed:
addendum: village fired attribute for single event firing

Cheaterpaul committed:
using categories within pools

Cheaterpaul committed:
efficient JigsawPiece check

Cheaterpaul committed:
added forgotten recalculation after adding new pieces

Cheaterpaul committed:
Converted weight to float and use new WeightedFloatRandom class

Cheaterpaul committed:
Clean changes

Cheaterpaul committed:
Print JigsawPatter changes to debug level

Cheaterpaul committed:
Select Category based on filename

Cheaterpaul committed:
Ability to add own categories for own JigsawPattern

Cheaterpaul committed:
1.15 port

Cheaterpaul committed:
comment mod 'custome_tnt_test' out in mods.toml

Cheaterpaul committed:
Merge branch '1.15.x' into 1.14_villagepools_4

Cheaterpaul committed:
Merge branch '1.15.x' into 1.14_villagepools_4

Cheaterpaul committed:
Add javadoc

Cheaterpaul committed:
remove obsolete class

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/ListJigsawPiece.java.patch:
```diff
@@ -0,0 +1,35 @@
+       } else {
+          this.field_214867_a = p_i51405_1_;
+          this.func_214864_b(p_i51405_2_);
++         this.resourceLocation = new net.minecraft.util.ResourceLocation(this.field_214867_a.stream().map(Object::toString).collect(Collectors.joining(", ")));
```
What are you even trying to create here?
Have you even looked at the resulting resource location?

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/SingleJigsawPiece.java.patch:
```diff
@@ -1,6 +1,23 @@
++
++   public SingleJigsawPiece(ResourceLocation location){
++      this(location, ImmutableList.of(), JigsawPattern.PlacementBehaviour.RIGID);
++   }
```
This doesn't seem necessary, as the String's are immediately converted to ResourceLocations so just specifying "modid:path" would work fine.

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/structure/PillagerOutpostPieces.java.patch:
```diff
@@ -0,0 +1,10 @@
+       JigsawManager.field_214891_a.func_214932_a(new JigsawPattern(new ResourceLocation("pillager_outpost/towers"), new ResourceLocation("empty"), ImmutableList.of(Pair.of(new ListJigsawPiece(ImmutableList.of(new SingleJigsawPiece("pillager_outpost/watchtower"), new SingleJigsawPiece("pillager_outpost/watchtower_overgrown", ImmutableList.of(new IntegrityProcessor(0.05F))))), 1)), JigsawPattern.P ...
+       JigsawManager.field_214891_a.func_214932_a(new JigsawPattern(new ResourceLocation("pillager_outpost/feature_plates"), new ResourceLocation("empty"), ImmutableList.of(Pair.of(new SingleJigsawPiece("pillager_outpost/feature_plate"), 1)), JigsawPattern.PlacementBehaviour.TERRAIN_MATCHING));
+       JigsawManager.field_214891_a.func_214932_a(new JigsawPattern(new ResourceLocation("pillager_outpost/features"), new ResourceLocation("empty"), ImmutableList.of(Pair.of(new SingleJigsawPiece("pillager_outpost/feature_cage1"), 1), Pair.of(new SingleJigsawPiece("pillager_outpost/feature_cage2"), 1), Pair.of(new SingleJigsawPiece("pillager_outpost/feature_logs"), 1), Pair.of(new SingleJigsawPi ...
++      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.JigsawPatternInitEvent.StructureJigsawPoolInitEvent.PillageOutpost());
```
This is in a static initalizer, and thus unpredictable.
The registry is already public so either we should take it over and fire a full registry event. Or let the end modders register things during their setup events.

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/structure/VillagePieces.java.patch:
```diff
@@ -0,0 +1,10 @@
+       SavannaVillagePools.func_214745_a();
+       DesertVillagePools.func_222739_a();
+       TaigaVillagePools.func_214806_a();
++      net.minecraftforge.event.world.JigsawPatternInitEvent.StructureJigsawPoolInitEvent.Village.fire();
```
Is none of the context needed? Yes, the vanilla pools dont need it. But that's because they are all static initializers.

LexManos commented on src/main/java/net/minecraftforge/event/world/JigsawPatternInitEvent.java:
```diff
@@ -0,0 +1,295 @@
+     * This event should be fired whenever a Jigsaw bases structure initializes its pools.
+     * Use this event to register custom JigsawPattern before the corresponding structure is constructed.
+     */
+    public static abstract class StructureJigsawPoolInitEvent extends Event
```
Super long internal class names defeat the entire point of internal classes.
Also, this isn't a child of the outer class.. why is it a inner class?

LexManos commented on src/main/java/net/minecraftforge/event/world/JigsawPatternInitEvent.java:
```diff
@@ -0,0 +1,295 @@
+     * @param category The category the building should be added. If null the default one will be selected. Some {@link JigsawPattern} may not contains the default category.
+     * @param pieces A list of JigsawPieces and weights that should be added to the pool
+     */
+    public void addBuildings(@Nonnull ResourceLocation category, @Nonnull List<Pair<JigsawPiece, Integer>> pieces)
```
Any reason why all these last parameters are lists instead of param arrays?

LexManos commented on patches/minecraft/net/minecraft/world/gen/feature/jigsaw/FeatureJigsawPiece.java.patch:
```diff
@@ -0,0 +1,34 @@
+    }
++
++   public net.minecraft.util.ResourceLocation getRegistryName() {
++      return registryName;
```
If you're gunna cache something at the beginning, and potentially get it wrong due to order of initalizers. Might as well optimize the toString as well.

LexManos reviewed (changes requested):
I am still sure if this is actually a problem or not. It could be argued that adding extra homes SHOULD decrease the rate in which farms spawn. As they are not considered seperate things. Nore am i seeing where exactly you properly weight things out. 
The entire jigsaw implementation, like everything else in vanilla, is hardcoded and annoying. But should be functional. This needs more discussion before moving on.

Cheaterpaul committed:
requested changes

- merge methods to param arrays
- move StructureJigsawPoolInitEvent  to separate class
- remove constructor with ResourceLocation in SingleJigsawPiece

Cheaterpaul committed:
Remove SingleJigsawPiece constructor with only ResourceLocation parameter
